crypto/signatures_test.go

```diff
@@ -1,25 +1,141 @@
 package crypto
 
 import (
+	"bytes"
 	"crypto/rand"
+	"errors"
 	"testing"
 
 	"github.com/NebulousLabs/Sia/encoding"
+	"github.com/NebulousLabs/ed25519"
 )
 
-// Creates and encodes a public key, and verifies that it decodes correctly,
-// does the same with a signature.
-func TestSignatureEncoding(t *testing.T) {
+// mockKeyDeriver is a mock implementation of keyDeriver that saves its provided
+// entropy and allows the client to specify the returned SecretKey and
+// PublicKey.
+type mockKeyDeriver struct {
+	called  bool
+	entropy [EntropySize]byte
+	sk      ed25519.SecretKey
+	pk      ed25519.PublicKey
+}
+
+func (kd *mockKeyDeriver) deriveKeyPair(entropy [EntropySize]byte) (ed25519.SecretKey, ed25519.PublicKey) {
+	kd.called = true
+	kd.entropy = entropy
+	return kd.sk, kd.pk
+}
+
+// Test that the Generate method is properly calling its dependencies and
+// returning the expected key pair.
+func TestGenerateRandomKeyPair(t *testing.T) {
+	var mockEntropy [EntropySize]byte
+	mockEntropy[0] = 5
+	mockEntropy[EntropySize-1] = 5
+	entropyReader := bytes.NewReader(mockEntropy[:])
+
+	sk := ed25519.SecretKey(&[SecretKeySize]byte{})
+	sk[0] = 7
+	sk[32] = 8
+	pk := ed25519.PublicKey(&[PublicKeySize]byte{})
+	pk[0] = sk[32]
+	kd := mockKeyDeriver{sk: sk, pk: pk}
+
+	// Create a SignatureKeyGenerator using mocks.
+	g := stdGenerator{entropyReader, &kd}
+
 	// Create key pair.
-	sk, pk, err := GenerateSignatureKeys()
+	skActual, pkActual, err := g.Generate()
+
+	// Verify that we got back the expected results.
 	if err != nil {
-		t.Fatal(err)
+		t.Error(err)
+	}
+	if *sk != skActual {
+		t.Errorf("Generated secret key does not match expected! expected = %v, actual = %v", sk, skActual)
```

Comment:
A shortcut I've taken frequently is to write a non-helpful error message when the test fails, with the intention of writing a more helpful error message if needed in the future.

90% of the checks you make in a test are never triggered, which is why I'm comfortable doing that. The other reason I'm comfortable doing that is because the context of the test is usually pretty obvious, and you have to look at the test code that failed anyway.

As a side effect, you can sometimes tell which tests I had trouble getting to pass because those tests will have much more verbose error messages.